crates/services/upgradable-executor/src/executor.rs

```diff
@@ -339,21 +346,70 @@ where
+        block: Components<TxSource>,
+        options: ExecutionOptions,
+    ) -> ExecutorResult<Uncommitted<ExecutionResult, Changes>>
+    where
+        TxSource: TransactionsSource + Send + Sync + 'static,
+    {
+        let block_version = block.header_to_produce.state_transition_bytecode_version;
+        let native_executor_version = self.native_executor_version();
+        if block_version == native_executor_version {
+            match &self.execution_strategy {
+                ExecutionStrategy::Native => self.native_produce_inner(block, options),
+                ExecutionStrategy::Wasm { module } => {
+                    self.wasm_produce_inner(module, block, options)
+                }
+            }
+        } else {
+            let module = self.get_module(block_version)?;
+            tracing::warn!(
+                "The block version({}) is different from the native executor version({}). \
+                The WASM executor will be used.", block_version, Self::VERSION
+            );
+            self.wasm_produce_inner(&module, block, options)
+        }
+    }
+
+    #[cfg(not(feature = "wasm-executor"))]
+    fn produce_inner<TxSource>(
+        &self,
+        block: Components<TxSource>,
+        options: ExecutionOptions,
+    ) -> ExecutorResult<Uncommitted<ExecutionResult, Changes>>
+    where
+        TxSource: TransactionsSource + Send + Sync + 'static,
+    {
+        let block_version = block.header_to_produce.state_transition_bytecode_version;
+        let native_executor_version = self.native_executor_version();
+        if block_version == native_executor_version {
+            self.native_produce_inner(block, options)
+        } else {
+            Err(ExecutorError::Other(format!(
+                "Not supported version `{block_version}`. Expected version is `{}`",
+                Self::VERSION
+            )))
+        }
+    }
+
+    #[cfg(feature = "wasm-executor")]
+    fn dry_run_inner<TxSource>(
         &self,
-        block: ExecutionBlockWithSource<TxSource>,
+        block: Components<TxSource>,
         options: ExecutionOptions,
     ) -> ExecutorResult<Uncommitted<ExecutionResult, Changes>>
     where
         TxSource: TransactionsSource + Send + Sync + 'static,
     {
-        let block_version = block.state_transition_version();
+        let block_version = block.header_to_produce.state_transition_bytecode_version;
         let native_executor_version = self.native_executor_version();
         if block_version == native_executor_version {
             match &self.execution_strategy {
-                ExecutionStrategy::Native => self.native_execute_inner(block, options),
+                ExecutionStrategy::Native => self.native_dry_run_inner(block, options),
                 ExecutionStrategy::Wasm { module } => {
-                    self.wasm_execute_inner(module, block, options)
+                    self.wasm_dry_run_inner(module, block, options)
                 }
             }
```

Comment:
Although this introduces some code duplication, the benefit is that it's simple to read and understand. However, I'm curious if you explored any alternative ways to implement these functions that may be a little DRYer. It is understandably challenging given the number of permuations dry run, validation, and Native/Wasm/wasm-executor enabled.

Reply:
Been looking at this. I think the only part I'd really want to DRY up more would be 
```rs
        let block_version = block.header_to_produce.state_transition_bytecode_version;
        let native_executor_version = self.native_executor_version();
        if block_version == native_executor_version { .. }
```
But honestly, that would save us like 1 line, especially since validate takes a different `Block` type than the other two, so to DRY all three, we'd have to get the version as a separate line.

I think it's already flat, and easy to read.

WRT to the functions doing essentially the same thing, I think they can all vary independently, so I don't think we should create an abstraction here.

Reply:
I was able to find a little place to DRY up the code a bit. Added `new_native_executor_instance` method.